src/ts/javaDistribution.ts

```diff
@@ -0,0 +1,164 @@
+ * Fires when the java distribution changes due to users changing it in workspace/user settings.
+ *
+ * Also fires on startup if there is a suitable Java distribution.
+ *
+ * Excludes Java versions that are incompatible.
+ */
+export const onDidChangeJavaDistribution = emitter.event;
+
+export type JavaDistribution = {
+  path: string;
+  version: number;
+};
+
+const resolveJava = async (path: string): Promise<JavaDistribution | null> => {
+  try {
+    const stats = await fs.stat(path);
+    if (!stats.isFile()) {
+      return null;
+    }
+    const result = await exec(`${path} -version`);
+    const { stderr } = result;
+    const versionStr = stderr.split('"')[1];
+    if (versionStr == null) {
+      console.warn(`Unexpected version string: ${stderr}`);
+      return null;
+    }
+    const majorVersion = versionStr.split(".")[0];
+    if (majorVersion == null) {
+      console.warn(`Malformed version: ${versionStr}`);
+      return null;
+    }
+    var version = parseInt(majorVersion);
+    return { path, version };
+  } catch (err: any) {
+    if (err.code !== "ENOENT") {
+      console.warn(`Received unexpected error when spawning java: ${err}`);
+    }
+    return null;
+  }
+};
+
+const extEnvVar = process.env.PATHEXT ?? "";
+const possibleJavaFilenames = extEnvVar.split(path.delimiter).map((it) => `java${it}`);
+
+const findJavaInDir = async (dir: string) => {
+  for (const filename of possibleJavaFilenames) {
+    const candidate = path.join(dir, filename);
+    const java = await resolveJava(candidate);
+    if (java != null) {
+      return java;
+    }
+  }
+  return null;
+};
+
+/**
+ * Find Java from either `$JAVA_HOME`, or `$PATH`.
+ */
+const findJavaFromSystem = async () => {
+  const javaHome = process.env.JAVA_HOME;
+  if (javaHome != null) {
+    var distro = await findJavaInDir(path.join(javaHome, "bin"));
+    if (distro != null && distro.version >= MINIMUM_JAVA_VERSION) {
+      emitter.fire(distro);
+      return;
+    }
+  }
+  const pathEnvVar = process.env.PATH;
+  if (pathEnvVar != null) {
+    for (const pathStr of pathEnvVar.split(path.delimiter)) {
+      const distro = await findJavaInDir(pathStr);
+      if (distro != null && distro.version >= MINIMUM_JAVA_VERSION) {
+        emitter.fire(distro);
+        return;
+      }
+    }
+  }
+  const response = await vscode.window.showInformationMessage(
+    `Cannot find suitable java in $PATH or $JAVA_HOME. pkl-lsp requires Java ${MINIMUM_JAVA_VERSION} or higher.`,
+    "Configure path to Java"
+  );
+  if (response === "Configure path to Java") {
+    vscode.commands.executeCommand(COMMAND_OPEN_WORKSPACE_SETTINGS, CONFIG_JAVA_PATH);
+  }
+};
+
+const handleConfiguredJavaPath = async (path: string) => {
+  const distribution = await resolveJava(path);
+  if (distribution === null) {
+    vscode.window
+      .showInformationMessage(
+        `Could not resolve java version information from ${config.javaPath}. Ensure it is the path to the java executable.`,
```

Comment:
```suggestion
        `Could not resolve Java version information from ${config.javaPath}. Ensure it is the path to the Java executable.`,
```
[throughout]

This error message doesn't tell the user where to configure this path. Is that obvious from the context where this error is displayed? (I lack VSCode context.)

Reply:
Yeah; basically, this message shows up after you've changed your workspace or user config. It should be pretty clear why this message is being shown to you.